book/src/dev/rfcs/xxxx-block-subsidy.md

```diff
@@ -0,0 +1,296 @@
+
+Inside `zebra-consensus/src/block/subsidy/` the following submodules will be created:
+
+- `general.rs`: General block reward functions and utilities.
+- `founders_reward.rs`: Specific functions related to funders reward.
+- `funding_streams.rs`: Specific functions for funding streams.
+
+In addition to calculations the block subsidy requires constants defined in the protocol. The implementation will also create additional constants, all of them will live at:
+
+- `zebra-consensus/src/parameters/subsidy.rs`
+
+Checking functions for blocks are implemented at `zebra-consensus/src/block/check.rs` and they are called at `zebra-consensus/src/block.rs`. This follows the already existing structure for block validation in Zebra.
+
+It is important to note that all blocks before Sapling are verified by the CheckpointVerifier, so they are not considered in this design. The following table will show what periods are included and what is not in this proposal:
+
+| Height                                 | Miner    | Founder reward | Funding streams | Shielded Coinbase | Target Spacing |
+|----------------------------------------|----------|----------------|-----------------|-------------------|----------------|             
+| ~Genesis~                              | ~**0%**~ | ~**0%**~       | ~**0%**~        | ~**No**~          | ~**None**~     |
+| ~(Genesis + 1)..Sapling~               | ~80%~    | ~20%~          | ~0%~            | ~No~              | ~150 seconds~  |
+| Sapling..Blossom                       | 80%      | 20%            | 0%              | No                | 150 seconds    |
+| Blossom..Heartwood                     | 80%      | 20%            | 0%              | No                | 75 seconds     |
+| Heartwood..Canopy                      | 80%      | 20%            | 0%              | Yes               | 75 seconds     |
+| Canopy..Second Halving                 | 80%      | 0%             | 20%             | Yes               | 75 seconds     |
+| Second Halving..                       | 100%     | 0%             | 0%              | Yes               | 75 seconds     |
+
+# Reference-level explanation
+[reference-level-explanation]: #reference-level-explanation
+
+Given the module structure proposed above, the block subsidy calculations from the protocol must be implemented. The final goal is to do semantic validation for each incoming block and make sure they pass the consensus rules.
+
+To do the calculations and checks the following constants, types and functions need to be introduced:
+
+## Constants
+
+### General
+
+- `SLOW_START_INTERVAL: Height`
+- `SLOW_START_SHIFT: Height`
+- `MAX_BLOCK_SUBSIDY: u64`
+- `BLOSSOM_POW_TARGET_SPACING_RATIO: u64`
+- `PRE_BLOSSOM_HALVING_INTERVAL: Height`
+- `POST_BLOSSOM_HALVING_INTERVAL: Height`
+
+### Founders Rewards
+
+- `FOUNDERS_FRACTION_DIVISOR: u64`
+- `FOUNDERS_ADDRESS_COUNT: u32`
+- `FOUNDER_ADDRESSES_MAINNET: [&str; 48]`
+- `FOUNDER_ADDRESSES_TESTNET: [&str; 48]`
```

Comment:
should this be `[char; 48]`?

edit: read further and saw that the funding stream receiver is actually 48 addresses, wow. Might be worth swapping out `&str` for a newtype that makes it clearer that there are 48 addresses in here.

edit2: is `zebra_chain::transparent:Address::PayToScriptHash` related here?

Reply:
I think this are `Address` types, we can do something as:

```
pub const FOUNDERS_REWARD_ADDRESSES_MAINNET: [Address; 48] = [
    Address::from_str("t3Vz22vK5z2LcKEdg16Yv4FFneEL1zg9ojd").expect("address is valid"),
    Address::from_str("t3cL9AucCajm3HXDhb5jBnJK2vapVoXsop3").expect("address is valid"),
    Address::from_str("t3fqvkzrrNaMcamkQMwAyHRjfDdM2xQvDTR").expect("address is valid"),
   ...
   ```

How do you see it ?

Reply:
If we checkpoint on Canopy, we won't need to implement this code at all.